io/fv3atm_oro_io.F90

```diff
@@ -0,0 +1,333 @@
+        !       Sfcprop(nb)%hprim(ix)     = oro%var2(i,j,1)
+        !--- hprime(1:14)
+        num = 1       ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num)  = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num) = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num) = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num) = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num) = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%hprime(ix,num) = oro%var2(i,j,num)
+        !--- oro
+        num = num + 1 ; Sfcprop(nb)%oro(ix)       = oro%var2(i,j,num)
+        num = num + 1 ; Sfcprop(nb)%oro_uf(ix)    = oro%var2(i,j,num)
+
+        Sfcprop(nb)%landfrac(ix)  = -9999.0
+        Sfcprop(nb)%lakefrac(ix)  = -9999.0
+
+        num = num + 1 ; Sfcprop(nb)%landfrac(ix)  = oro%var2(i,j,num) !land frac [0:1]
+        if (Model%lkm > 0  ) then
+          if(oro%var2(i,j,num+1)>Model%lakefrac_threshold .and. &
+               oro%var2(i,j,num+2)>Model%lakedepth_threshold) then
+            Sfcprop(nb)%lakefrac(ix)  = oro%var2(i,j,num+1) !lake frac [0:1]
+            Sfcprop(nb)%lakedepth(ix) = oro%var2(i,j,num+2) !lake depth [m]    !YWu
+          else
+            Sfcprop(nb)%lakefrac(ix)  = 0
+            Sfcprop(nb)%lakedepth(ix) = -9999
+          endif
+        else
+          Sfcprop(nb)%lakefrac(ix)  = oro%var2(i,j,num+1) !lake frac [0:1]
+          Sfcprop(nb)%lakedepth(ix) = oro%var2(i,j,num+2) !lake depth [m]    !YWu
+        endif
+        num = num + 2 ! To account for lakefrac and lakedepth
+
+        Sfcprop(nb)%vegtype_frac(ix,:)  =  -9999.0
+        Sfcprop(nb)%soiltype_frac(ix,:) =  -9999.0
+
+        Sfcprop(nb)%vegtype_frac(ix,:)  = oro%var3v(i,j,:) ! vegetation type fractions, [0:1]
+        Sfcprop(nb)%soiltype_frac(ix,:) = oro%var3s(i,j,:) ! soil type fractions, [0:1]
+
+      enddo
+    enddo
+
+    !--- deallocate containers and free restart container
+    deallocate(oro%name2)
+    deallocate(oro%var2)
+    deallocate(oro%var3v)
+    deallocate(oro%var3s)
+
+    nullify(oro%name2)
+    nullify(oro%var2)
+    nullify(oro%var3v)
+    nullify(oro%var3s)
+
+  end subroutine Oro_io_copy
+
+  !>@brief Destructor for Oro_io_data_type
+  subroutine Oro_io_final(oro)
+    implicit none
+    type(Oro_io_data_type) :: oro
+
+    ! This #define reduces code length by a lot
+#define IF_ASSOC_DEALLOC_NULL(var) \
```

Comment:
Just a quick question for my understanding. This works because `name`, `var2`, ... are all pointers? Would it have been possible to create a subroutine instead to achieve the same (sorry, came back from Japan last night, my brain is not firing on all cylinders this morning).

Reply:
To use a subroutine, you would need to pass a reference to a pointer, which is impossible in Fortran. Also, speaking of this construct in general, there are multiple types of pointers to be deallocated. Thus, you would need multiple subroutines. Even, hypothetically, if you could use subroutines, it would be a whole lot shorter to use a #define.

Reply:
Thanks for the explanation. I am fine with your solution!

Reply:
In Fortran arguments are passed by reference, by default.

Reply:
I don't think that is true of pointers. If you assign to a pointer, you assign to a local copy, not the one that the caller passes in. In other words, the pointer itself is the reference. You need the equivalent of this:

```C
void my_c_function(int **pointer_to_a_pointer)
{
    if(*pointer_to_a_pointer)
    {
       free(*pointer_to_a_pointer);
       *pointer_to_a_pointer = nullptr;
    }
}
```

Reply:
This is undergoing final testing now. I think it's too late to change it. We should wait to a later clean-up PR, and put something in fv3atm_common_io.F90.

Reply:
Off course. I am not even sure if we should make any change at all. Since, as you said, subroutine will need to be overloaded.

Reply:
OK, just making sure all is good here before we start testing. I assume it is and will resume testing.

Reply:
@SamuelTrahanNOAA if all good here, can you resolve this conversation , please?